### PR TITLE
Added union support to auto middleware.

### PR DESCRIPTION
**Pull request by richard-powers**
- Should we add e2e tests?
- Any other testing?

### CODE REVIEW

File: src/middleware/auto.zig [hunk 1]
```diff
@@ -143,56 +436,49 @@ fn analyze(comptime Context: type, comptime stack: []const type) []const type {
...
+    return switch (@typeInfo(T)) {
+        .optional => |o| o.child,
+        else => T,
     };
-    const M3 = struct {
-        const D = struct {
-            m: M1,
-            m2: *M2,
-        };
+}
```
Comment (BTOdell) [inline src/middleware/auto.zig]: Pretty sure this is what `Unwrap` does.

File: src/middleware/auto.zig [hunk 2]
```diff
@@ -238,17 +528,35 @@ fn field(comptime Context: type, comptime Middleware: type) Type.StructField {
...
+                if (@typeInfo(vf.type) == .@"struct") {
+                    for (@typeInfo(vf.type).@"struct".fields) |sf| {
+                        if (Unwrap(sf.type) == Middleware) break :blk true;
+                    }
+                }
+            }
+        }
+        break :blk false;
+    };
+}
```
Comment (BTOdell) [inline src/middleware/auto.zig:566]: Adding this function here separates the `field` function from its tests below.

File: src/middleware/auto.zig [hunk 3]
```diff
@@ -111,16 +388,32 @@ fn extractDependency(comptime T: type, comptime top_level: bool) type {
...
+/// Skipped when scanning a context so their types never reach extractDependency,
+/// which matters because a special field may be a pointer (otherwise rejected).
+const reserved_field_names = [_][]const u8{ "query_params", "body", "req" };
+
+fn isReservedField(comptime name: []const u8) bool {
+    for (reserved_field_names) |reserved| {
+        if (std.mem.eql(u8, name, reserved)) return true;
+    }
+    return false;
 }
```
Comment (BTOdell) [inline src/middleware/auto.zig]: This could use a static string map to be faster and not require a function.

File: src/middleware/auto.zig [hunk 4]
```diff
@@ -37,67 +40,341 @@ pub fn auto(comptime Context: type, ctx: *MiddlewareContext(Context)) !void {
...
+}
+
+/// Whether Context has a guaranteed (non-union) field whose type unwraps to T.
+fn hasFieldOfType(comptime Context: type, comptime T: type) bool {
+    inline for (@typeInfo(Context).@"struct".fields) |f| {
+        if (@typeInfo(f.type) == .@"union") continue;
+        if (Unwrap(f.type) == T) return true;
+    }
+    return false;
+}
```
Comment (BTOdell) [inline src/middleware/auto.zig:346]: This seems similar to the `extractMiddleware` function.

File: src/middleware/auto.zig [hunk 5]
```diff
@@ -37,67 +40,341 @@ pub fn auto(comptime Context: type, ctx: *MiddlewareContext(Context)) !void {
...
+
+/// Whether the guaranteed (non-union) field of `Context` whose type unwraps to `M` is declared optional.
+/// Returns null when `Context` has no such field.
+fn fieldProvidingIsOptional(comptime Context: type, comptime M: type) ?bool {
+    inline for (@typeInfo(Context).@"struct".fields) |f| {
+        if (@typeInfo(f.type) == .@"union") continue;
+        if (Unwrap(f.type) == M) return @typeInfo(f.type) == .optional;
+    }
+    return null;
+}
```
Comment (BTOdell) [inline src/middleware/auto.zig]: This also seems similar to `extractMiddleware`.

**Review by BTOdell — CHANGES_REQUESTED** [review]

File: src/middleware/auto.zig (hunk 1 above)
Comment (richard-powers) [reply src/middleware/auto.zig]: The behavior is different

Comment (BTOdell) [reply src/middleware/auto.zig]: Just because `Unwrap` is recursive? Or because it handles our Optional type? Besides those two things, it does the same thing.

Comment (richard-powers) [reply src/middleware/auto.zig]: Right, it doesn't unwrap the custom `Optional(T)` type intentionally

File: src/middleware/auto.zig (hunk 4 above)
Comment (richard-powers) [reply src/middleware/auto.zig:346]: I don't really see how they're similar, are you seeing a way we could combine them somehow?

File: src/middleware/auto.zig (hunk 1 above)
Comment (richard-powers) [reply src/middleware/auto.zig]: I thought I had created it for some specific edge case but the tests are passing just using `Unwrap` so I'll change it back

File: src/middleware/auto.zig (hunk 4 above)
Comment (BTOdell) [reply src/middleware/auto.zig:346]: Sorry, I meant the `field(...)` function, not `extractMiddleware`. GitHub diff collapsing confused me lol

Comment (richard-powers) [reply src/middleware/auto.zig:346]: Gotcha, updated

**Review by BTOdell — APPROVED** [review]
